roachprod: create load balancer (gcloud)

herkolategan opened the pull request:
Previously, we did not support load balancers and workloads usually connected to
a single node with fallback support to other nodes. This change adds the ability
to create a load balancer for a managed cluster (only supports GCE currently).

A load balancer can be created for each service (virtual cluster). Load
balancers are discerned by the port they serve. A simple health check is also
added to monitor the connectivity on these ports. Currently, the load balancers
will be set to use utilisation as a metric for load balancing. But if required
we can switch this to something else.

Load balancers consist of various components in Google Cloud that work together
to achieve load balancing. The main parts require an instance group that becomes
part of a backend service. Forwarding rules are then set up to configure the
correct ports and policies.

See: #119009
Epic: CRDB-33832

Release Note: None

herkolategan committed:
roachprod: create load balancer (gcloud)

Previously we did not support load balancers and workloads usually connected to
a single node with fallback support to other nodes. This change adds the ability
to create a load balancer for a managed cluster (only support GCE currently).

A load balancer can be created for each service (virtual cluster). Load
balancers are discerned by the port they serve. A simple health check is also
added to monitor the connectivity on these ports. Currently, the load balancers
will be set to use utilisation as a metric for load balancing. But if required
we can switch this to something else.

Load balancers consist of various components in Google Cloud that work together
to achieve load balancing. The main parts require an instance group that becomes
part of a backend service. Forwarding rules are then set up to configure the
correct ports and policies.

See: #119009
Epic: CRDB-33832

Release Note: None

herkolategan committed:
roachprod: destroy load balancer resources

Adding a load balancer creates several resources in Google Cloud. This change
ensures these resources are destroyed when the cluster is destroyed.

Some resources are only implicitly linked to a load balancer or managed instance
group and could dangle if the delete command fails halfway through. Health
checks are one such resource, hence here we rely on the naming scheme to find
these resources, rather than references via other components.

See: #119009
Epic: CRDB-33832

Release Note: None

herkolategan committed:
roachprod: add load balancer to provider interface

Expose the load balancer functionality through the provider interface, so that
it can be used via the `roachprod` CLI or by `roachtest`.

See: #119009
Epic: CRDB-33832

Release Note: None

herkolategan committed:
roachprod: create load balancer command

Add the create load balancer command to the `roachprod` CLI. The
`CreateLoadBalancer` function is also exposed as part of the `roachprod` API.
This function will create a load balancer given a virtual cluster name
(service).

The service needs to use the same port across the whole cluster for load
balancing to function correctly, otherwise health checks will fail, in addition
to connection problems.

A load balancer will be created for the port the service runs on, and the
resource names will also include the port name to discern between various load
balancers. It's possible to create more than one load balancer per cluster, if
the cluster is running multiple services.

See: #119009
Epic: CRDB-33832

Release Note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/119382)
<!-- Reviewable:end -->

herkolategan commented:
Note: I still want to add better progress / status updates (it's pretty bad on this one, plenty of waiting in silence), but I plan to introduce it in a follow-up PR for both load balancer & grow commands. If need be I can do it in this PR.

renatolabs reviewed (commented):
Nice, excited about this change!

I want to play with this locally, I'll make some time for it tomorrow. In the mean time, I'm curious: why do we need the `load-balance` command? I was hoping that, in practice, all managed clusters would have a load balancer out of the box.

Or maybe we could make that the default for roachtest but not for roachprod.

I'm also wondering if a `--load-balance` option to `create` is better than requiring people to `create` + `load-balance`.

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 5 of 5 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119382#-:-NrpzKy_4ujZPPbBPATq:b-scrlox)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong and @herkolategan)

___
*[`pkg/roachprod/roachprod.go` line 2412 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119382#-Nrpz4TS4HyQkaNfyVAB:-Nrpz4TS4HyQkaNfyVAC:b-s8a3z0) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfa0e0f5a7f36a3d7b7bfbafc51fea4e5cc7930f/pkg/roachprod/roachprod.go#L2412)):*
> ```Go
> 	port := services[0].Port
> 	for _, service := range services[1:] {
> 		if port != service.Port {
> ```

This should hold by construction on every roachprod cluster, yes? It would require a fair bit of manual effort for this not to be the case, as far as I understand.
___
*[`pkg/roachprod/vm/azure/azure.go` line 146 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119382#-NrpyGqq68_tEopIcKf1:-NrpyGqq68_tEopIcKf2:b-os4ljq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab04327cf7f91ee5cc74d4af4310c0736e54555b/pkg/roachprod/vm/azure/azure.go#L146)):*
> ```Go
> 
> func (p *Provider) CreateLoadBalancer(*logger.Logger, vm.List, int) error {
> 	panic("unimplemented")
> ```

I know you are just following the existing pattern, so I'm just thinking out loud: these really should just return an error. In the off chance we call this during a nightly, we just want the cluster creation to fail and move on, no need to crash the entire test run.
___
*[`pkg/roachprod/vm/gce/gcloud.go` line 2060 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119382#-NrpxhQf6Rta9WKw1xgE:-NrpxhQg39WPstKPxpTo:b-z8wqor) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c601e729ffd6c0017708d133f6fcc89c571941f/pkg/roachprod/vm/gce/gcloud.go#L2060)):*
> ```Go
> 	var g errgroup.Group
> 	for cluster, project := range clusterProjectMap {
> 		cluster, project := cluster, project // capture loop variables
> ```

Was this broken before? I don't see the capture in the code added in this commit.


<!-- Sent from Reviewable.io -->

herkolategan reviewed (commented):
Thanks for the review / comments!

> why do we need the `load-balance` command
It adds quite a delay to creation (and deletion), and has extra resource costs. And you need to configure the ports which means you can't technically have a load balancer till you have a service (but we could maybe create one for the first / storage port in advance?) For all these reasons I opted to make it optional and per service. I think there is a way to load balance a range of ports, but a load balancer (or backend rather) only allows for one health check (one port) currently. So in that regard if we want health checks for all services, it's a load balancer per service. Happy for feedback on this, and maybe to try out other configurations (so many knobs on this tbh).

> Or maybe we could make that the default for roachtest but not for roachprod.
This could be a viable option.

> I'm also wondering if a `--load-balance` option to `create` is better than requiring people to `create` + `load-balance`.
This comes down to requiring the service (port) for the health check, and if you wanted one specific to a tenant / virtual cluster you would need to be able to specify it afterwards. Without health checks we could probably do a range upfront. I can think about this for a bit unless you have some ideas.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119382#-:-NrtQgGKDwIK9Ln4rVa_:b-5osr92)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong and @renatolabs)

___
*[`pkg/roachprod/roachprod.go` line 2412 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119382#-Nrpz4TS4HyQkaNfyVAB:-NrtRi6001wzQIeax252:b-rz9xc8) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfa0e0f5a7f36a3d7b7bfbafc51fea4e5cc7930f/pkg/roachprod/roachprod.go#L2412)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

This should hold by construction on every roachprod cluster, yes? It would require a fair bit of manual effort for this not to be the case, as far as I understand.
</blockquote></details>

Typically yes, there would have to be some weird port skew on a specific node. Or the case which I was thinking of is if someone started tenants per-node on different ports, which is easily achievable considering we can pass the --port flag now. Not sure if someone would have a reason to do so.
___
*[`pkg/roachprod/vm/azure/azure.go` line 146 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119382#-NrpyGqq68_tEopIcKf1:-NrtSBcl6A7nwff4MfwW:b19mupl) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab04327cf7f91ee5cc74d4af4310c0736e54555b/pkg/roachprod/vm/azure/azure.go#L146)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

I know you are just following the existing pattern, so I'm just thinking out loud: these really should just return an error. In the off chance we call this during a nightly, we just want the cluster creation to fail and move on, no need to crash the entire test run.
</blockquote></details>

Makes sense, maybe a new issue to handle all of the unsupported commands for different providers to prevent a panic? I'll create one.
___
*[`pkg/roachprod/vm/gce/gcloud.go` line 2060 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119382#-NrpxhQf6Rta9WKw1xgE:-NrtkwHH-mmdl6px5qyb:bwkosf5) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c601e729ffd6c0017708d133f6fcc89c571941f/pkg/roachprod/vm/gce/gcloud.go#L2060)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Was this broken before? I don't see the capture in the code added in this commit.
</blockquote></details>

Nope this was a knee-jerk reaction to passing them into a function... face-palm., which obviously doesn't require it. I'll remove these.


<!-- Sent from Reviewable.io -->

herkolategan reviewed (commented):
> it failed because I didn't have the `beta` gcloud component installed.

Thanks for also testing! Good point around failure not being recoverable, I'll add a destroy-catch if the load-balancer fails to create. Or I'm considering adding a destroy load-balancer type command as well. Part of me is tempted to do it like you mentioned where load-balancers are implicitly created, or via flag. But I think that should be left for another PR once it's stable.

> silent for about a minute

Yes, as mentioned above I want to do this in another PR for both Grow and load balancers.

> do we need `gc` updates to deal with dangling resources

I tested the destroy part a few times. The resources themselves are dependent on one another, so delete load balancer will be able to clean up partial situations. The only thing that can dangle are the health-checks, so I added a check in delete for those.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119382#-:-NsN6n3GEOg__dFLUoHr:b-w3kwmt)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong and @renatolabs)



<!-- Sent from Reviewable.io -->

herkolategan reviewed (commented):
Forgot to mention, I realised `--secure` does not work with this. So this PR is a bit in flux, I'm going to add another commit to cater for secure clusters.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119382#-:-NsN7wvQAKvSM2pv3AZM:bx2w9sv)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong and @renatolabs)



<!-- Sent from Reviewable.io -->

herkolategan committed:
roachprod: clean-up partially created load balancer

In the event that load balancer creation fails, any resources that did manage to
be created should be cleaned up. This change exposes deleting a load balancer
through the provider interface. Deletion does not require any resources to
exists, but it will find any associated resources by name and clean it up.

See: #119009
Epic: CRDB-33832

Release Note: None

herkolategan committed:
roachprod: add load balancer IP to certs

The Load Balancer IP needs to be added to the cluster certificates, either when
starting new nodes. Or when creating a new load balancer. We only need to
regenerate one of the certificates `node` and redistribute this throughout the
cluster. After making changes to the certificates a signal `SIGHUP` can be sent
to the cockroach processes in the cluster to cause a reload of the certificates.

See: #119009
Epic: CRDB-33832

Release Note: None

renatolabs reviewed (commented):
Flushing some comments while I experiment with the new commands locally.

renatolabs reviewed (approved):
Used this locally and I think it works well. Discussed with Herko offline on certain limitations or assumptions of the current implementation that I think should be more explicit or documented somewhere.

The comments I left earlier can also be done in future work, so I'm approving! Nice work!

herkolategan commented on pkg/roachprod/roachprod.go:
```diff
@@ -2376,3 +2376,48 @@ func createAttachMountVolumes(
+	clusterName string,
+	secure bool,
+	virtualClusterName string,
+	sqlInstance int,
```
`sql instance` is the offset on a VM itself, so if we ever have a deployment of two tenant (of the same tenant) instances per node, you'll want to specify the sql instance offset.

The `cli` should be passing this though. Thanks for catching, I'll update that.

renatolabs replied:
Oh, I see what you mean. So, effectively, only one of the instances in the node can be part of the load balancing. 

(Related aside: I wonder if there's ever a use-case to start multiple instances of the same tenant on the same node, especially since we don't provide any kind of resource isolation. We might have been able to do away with "sql instances" altogether. But that's a different conversation.)

herkolategan replied:
@renatolabs I'm all for removing this. Add unnecessary complexity, and not sure if we gain any coverage from testing it. I remember the other option was to hide it in a struct, but that only slightly reduces its footprint.

herkolategan commented on pkg/roachprod/install/cluster_synced.go:
```diff
@@ -1843,7 +1886,15 @@ func (c *SyncedCluster) createNodeCertArguments() ([]string, error) {
+		return nil, err
+	}
+	for _, lb := range lbAddresses {
+		fmt.Println("Appending load balancer IP", lb.IP)
```
Nope, this was debugging. Nice catch.

herkolategan commented on pkg/roachprod/vm/gce/gcloud.go:
```diff
@@ -1910,6 +1910,37 @@ func (p *Provider) CreateLoadBalancer(_ *logger.Logger, vms vm.List, port int) e
+		if !ok {
+			continue
+		}
+		if cluster == clusterName && resourceType == "forwarding-rule" {
```
True, but since other people are also creating forwarding rules outside of `roachprod` it's a bit of an extra sanity check to make sure we're not dealing with a `non-roachprod` forwarding rule. Which is also checked by the `roachprod` tag in the name.

herkolategan committed:
roachprod: make virtual cluster optional for load balancer

See: #119009
Epic: CRDB-33832

Release Note: None

herkolategan commented:
TFTR!

bors r=renatolabs
